gc/2022/GC-08-23.md

```diff
@@ -33,3 +33,187 @@ Installation is required, see the calendar invite.
+
+BT: I’ll +1 to that, the tools want annotations and we might want to keep the annotations indefinitely, but have the tools do the compression. 
+
+DG: That’s a reasonable answer, that assumes that we do the work to handle compression at a different layer, and that the tools do the right thing and we have the tools in place to do that right?
+
+BT: Right, Isee layer 1 compression as the thing that’ll give us the best of both worlds. We could use the Wasm type system as the type compression, and still have type annotation in the bytecode for future tools, and future feature extensions
+
+DG: Any other thoughts? 
+
+CW: Restate something came up implicitly, in a couple of years from now, if we do have a more solidified view of what extensions we want in the future, we can still add annotationless versions in the future with essentially the same code size savings. 
+
+AR: That’s the conservativeness argument
+
+DG: It doesn’t have to be an example for this particular set, but is there an example that you could point to of something that was broken, looking for something that’s not as hypothetical as future problems, extensions, but even a toy type system example that can run into this problem
+
+AR: Select instruction is an example it didn’t have annotations, and once we added subtyping we realized we have to compute LUBs now we did this hack, where it was moderately ugly, we’ve already run into this problem, the br_table was also simple but hack not so ugly, simple example that everyone can understand. With the kind of things we’re talking about now, the problems would be much more intricate. IT’s not completely generic types, it’s very specific types
+
+FM: Is it fair to say that you run into trouble when you have to do the LUB?
+
+AR: Depends on how you define trouble, it’s very expensive, depending on how large your types are. In Wasm, we require that all types that occur in validation are defined, when you compute an LUB you may have a type that doesn’t have a definition, doesn’t have an index it can refer to.  You have to extend the whole type system with synthesized types, that’s a whole new level of complexity as well which we do avoid right now
+
+BT: Ran into this problem for func.bind, which has a synthesized type, we run into the same issue with any type of inference that it creates types that you haven’t written down. It’s because we have this subtyping relationship that we have to declare, if you haven’t written one of the subtypes in the chain, it’s essentially a synthesized type, and how is that type canonicalized? We don’t have to mandate type canonicalization
+
+AR: LUBs only work when you have proper lattice in the type system, or at least a semi lattice. For br_table you need GLBs (?) so you would need a full lattice, that’s a severe constraint. There are cases where you can’t synthesize a proper type, there is no principle type that you could synthesize, it’s not clear how you would solve this problem
```

Comment:
@rossberg, why would br_table need GLBs?

Reply:
@tlively, this was a problem we ran into when we introduced reference types and that we could only fix by adding bottom, see my presentation from June 2019: https://github.com/WebAssembly/meetings/blob/main/main/2019/presentations/2019-06-rossberg-reference-types.pdf

FWIW, people keep being surprised by this one, see e.g. [this](https://github.com/WebAssembly/spec/pull/1521#issuecomment-1220548348) very recent comment.

Reply:
Thanks, that makes sense.